prov/shm/src/smr_domain.c

```diff
@@ -69,6 +71,128 @@ static struct fi_ops smr_domain_fi_ops = {
+{
+	struct smr_domain *domain;
+	struct smr_mr_entry *mr;
+	int ret = 0;
+
+	mr = container_of(fid, struct smr_mr_entry, mr_fid.fid);
+	domain = mr->domain;
+
+	fastlock_acquire(&domain->util_domain.lock);
+	ret = ofi_mr_map_remove(&domain->mr_map, mr->key);
+	fastlock_release(&domain->util_domain.lock);
+	if (ret)
+		return ret;
+
+	ofi_atomic_dec32(&domain->util_domain.ref);
+	free(mr);
+	return 0;
+}
+
+static struct fi_ops smr_mr_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = smr_mr_close,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static int smr_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
+			  uint64_t flags, struct fid_mr **mr)
+{
+	struct smr_domain *domain;
+	struct smr_mr_entry *mr_entry;
+	uint64_t key;
+	int ret = 0;
+
+	if (fid->fclass != FI_CLASS_DOMAIN || !attr || attr->iov_count <= 0)
+		return -FI_EINVAL;
+
+	domain = container_of(fid, struct smr_domain, util_domain.domain_fid.fid);
+	mr_entry = calloc(1, sizeof(*mr_entry));
+	if (!mr_entry)
+		return -FI_ENOMEM;
+
+	fastlock_acquire(&domain->util_domain.lock);
+
+	mr_entry->mr_fid.fid.fclass = FI_CLASS_MR;
+	mr_entry->mr_fid.fid.context = attr->context;
+	mr_entry->mr_fid.fid.ops = &smr_mr_fi_ops;
+	mr_entry->domain = domain;
+	mr_entry->flags = flags;
+
+	ret = ofi_mr_map_insert(&domain->mr_map, attr, &key, mr_entry);
+	if (ret) {
+		free(mr_entry);
+		goto out;
+	}
+
+	mr_entry->mr_fid.key = mr_entry->key = key;
+	mr_entry->mr_fid.mem_desc = (void *) (uintptr_t) key;
+
+	*mr = &mr_entry->mr_fid;
+	ofi_atomic_inc32(&domain->util_domain.ref);
+
+out:
+	fastlock_release(&domain->util_domain.lock);
+	return ret;
+}
+
+static int smr_mr_regv(struct fid *fid, const struct iovec *iov,
+		       size_t count, uint64_t access, uint64_t offset,
+		       uint64_t requested_key, uint64_t flags,
+		       struct fid_mr **mr, void *context)
+{
+	struct fi_mr_attr attr;
+
+	attr.mr_iov = iov;
+	attr.iov_count = count;
+	attr.access = access;
+	attr.offset = offset;
+	attr.requested_key = requested_key;
+	attr.context = context;
+	return smr_mr_regattr(fid, &attr, flags, mr);
+}
+
+static int smr_mr_reg(struct fid *fid, const void *buf, size_t len,
+		      uint64_t access, uint64_t offset, uint64_t requested_key,
+		      uint64_t flags, struct fid_mr **mr, void *context)
+{
+	struct iovec iov;
+
+	iov.iov_base = (void *) buf;
+	iov.iov_len = len;
+	return smr_mr_regv(fid, &iov, 1, access, offset, requested_key, flags,
+			   mr, context);
+}
+
+static struct fi_ops_mr smr_mr_ops = {
+	.size = sizeof(struct fi_ops_mr),
+	.reg = smr_mr_reg,
+	.regv = smr_mr_regv,
+	.regattr = smr_mr_regattr,
+};
+
+int smr_mr_verify(struct smr_domain *smr_domain, ssize_t len,
+		  uint64_t addr, uint64_t key, uint64_t access)
+{
+	int ret;
+
+	fastlock_acquire(&smr_domain->util_domain.lock);
+	ret = ofi_mr_map_verify(&smr_domain->mr_map, (uintptr_t *) &addr, len,
+				key, access, NULL);
+	fastlock_release(&smr_domain->util_domain.lock);
+	return ret;
+}
```

Comment:
The MR code is generic.  I would add this to the util provider.  The tcp provider should be able to use the same implementation.

Reply:
How exactly would you like to incorporate it into the util provider? The ofi_mr_map_verify is already the util call. The only thing missing is the domain locking but it seemed like the utility code was trying to stay away from locking and assuming that locking had already happened. Were you thinking to put it into in util_domain, taking in the util_domain and mr_map, locking it, and calling ofi_mr_map_verify through there?

Reply:
I was referring to most of the changes to this file (from this commit) -- smr_mr_entry, smr_mr_close, smr_mr_regattr/regv/reg, etc.